docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,1259 @@
+spec:
+  controlPlane:
+    highAvailability:
+      enabled: true # by default HA is disabled.
+      failureDomain:
+        type: <single-zone | multi-zone>
+        whenUnsatisfied: <DoNotSchedule | ScheduleAnyway>
+```
+
+An additional capability is provided to the consumer to clearly define the behavior if the chosen type of failure domain cannot be provisioned by gardener due lack of available seeds capable of either hosting a `single-zone` or `multi-zone` shoot control plane.
+
+### Future enhancements
+
+Following additional options influencing scheduling and provisioning of shoot control plane components could be provided in the future:
+
+_Failure tolerance_
+
+In the context of an etcd cluster, a 3 member cluster will only provide a failure tolerance of 1 availability zone (for multi-zone) or node (for single-zone-multi-node). This can still be the default setup, but an option could be provided to make it configurable.
+
+```yaml
+kind: Shoot
+apiVersion: core.gardener.cloud/v1beta1
+spec:
+  controlPlane:
+    highAvailability:
+      enabled: true # by default HA is disabled.
+      type: <single-zone | multi-zone>
+      failureTolerance: <possible values should be in the range 1..2(included)>
+```
+
+## Gardener Scheduler
+
+A scheduling request could be for a shoot with HA or non-HA control plane. It is therefore required to appropriately select a seed. There are two use cases that need to be considered:
+
+**Case #1**
+
+A shoot which has no seed assigned is to be scheduled.
+
+_Proposed Changes_
+
+Introduce a new filtering step in [reconciler](https://github.com/gardener/gardener/blob/4dbba56e24b5999fd728cb84b09a4d6d54f64479/pkg/scheduler/controller/shoot/reconciler.go#L110) which will do the following:
+
+* Segregate available seeds into `multi-zonal` seeds and `non-multi-zonal` seeds.
+* Assign selection priority based on shoot spec HA configuration for the control plane.
+  * Shoot with `multi-zonal` control plane can only be scheduled on `multi-zonal` seed. Therefore for these shoots the target list of seeds will only be the ones which are `multi-zonal`.
+  * Shoot with `single-zonal` control plane can be scheduled either on `single-zonal` seed or `multi-zonal` seeds. In order to find the best fit - first check if there is any available seed in `single-zonal` seeds. If there is one, then that is assigned. If there is no available seed in `non-multi-zonal` seeds, then look for an available seed in `multi-zonal` seeds.
+
+The above selection algorithm should be applied only after applying the [strategy](https://github.com/gardener/gardener/blob/4e2d28f2af7093eb94819652a6f4709b5fbfaf06/pkg/scheduler/controller/shoot/reconciler.go#L256-L274). This ensures that strategy based selection e.g `region affinity` or `minimum distance` is applied first and then selection of a `single-zone` or `multi-zone` seed cluster is made to optimize the geographic separation between the shoot cluster and the shoot cluster control plane.
```

Comment:
How can it be applied "afterwards"? Filtering happens always beforehand. You first decide which seeds are eligible. Then you calculate the score.

Reply:
Ok i see that there is a problem in the way it has been put. The original intent was the following:
1. multi-zonal shoot control planes can only be scheduled on multi-zonal seeds. This is clear. The filtering happens early for that.
2. single-zonal shoot control planes can be hosted on both multi-zonal and non-multi-zonal seeds with preference given to non-multi-zonal shoots. Therefore this list will be sorted and arranged in a manner where non-multi-zonal seeds will be put before multi-zonal seeds.
Once this filtering is done then strategy is applied (current strategies that are in place are minimal-distance and same-region.

Reply:
Please have a look at the revised section.